R/hardcode_no_ct.R

```diff
@@ -0,0 +1,69 @@
+#' Derive an SDTM variable with a hardcoded value
+#'
+#' [hardcode_no_ct()] maps a hardcoded value to a target SDTM variable that has
+#' no terminology restrictions.
+#'
+#' @param raw_dataset The raw dataset.
+#' @param raw_variable The raw variable.
+#' @param target_sdtm_variable The target SDTM variable.
+#' @param target_hardcoded_value Hardcoded value.
+#' @param target_dataset Target dataset. By default the same as `raw_dataset`.
+#' @param merge_to_topic_by If `target_dataset` is different than `raw_dataset`,
+#'   then this parameter defines keys to use in the join between `raw_dataset`
+#'   and `target_dataset`.
+#'
+#' @examples
+#' MD1 <-
+#'   tibble::tribble(
+#'     ~oak_id, ~raw_source, ~patient_number, ~MDRAW,
+#'     1L, "MD1", "PATNUM", "BABY ASPIRIN",
+#'     2L, "MD1", "PATNUM", "CORTISPORIN",
+#'     3L, "MD1", "PATNUM", NA_character_,
+#'     4L, "MD1", "PATNUM", "DIPHENHYDRAMINE HCL"
+#'   )
+#'
+#' # Derive a new variable `CMCAT` by overwriting `MDRAW` with the
+#' # hardcoded value "GENERAL CONCOMITANT MEDICATIONS".
+#' hardcode_no_ct(
+#'   raw_dataset = MD1,
+#'   raw_variable = "MDRAW",
+#'   target_sdtm_variable = "CMCAT",
+#'   target_hardcoded_value = "GENERAL CONCOMITANT MEDICATIONS"
+#' )
+#'
+#' CM_INTER <-
+#'   tibble::tribble(
+#'     ~oak_id, ~raw_source, ~patient_number, ~CMTRT, ~CMINDC,
+#'     1L, "MD1", "PATNUM", "BABY ASPIRIN", NA,
+#'     2L, "MD1", "PATNUM", "CORTISPORIN", "NAUSEA",
+#'     3L, "MD1", "PATNUM", "ASPIRIN", "ANEMIA",
+#'     4L, "MD1", "PATNUM", "DIPHENHYDRAMINE HCL", "NAUSEA",
+#'     5L, "MD1", "PATNUM", "PARACETAMOL", "PYREXIA"
+#'   )
+#'
+#' # Derive a new variable `CMCAT` by overwriting `MDRAW` with the
+#' # hardcoded value "GENERAL CONCOMITANT MEDICATIONS" with a prior join to
+#' # `target_dataset`.
+#'
+#' hardcode_no_ct(
+#'   raw_dataset = MD1,
+#'   raw_variable = "MDRAW",
+#'   target_sdtm_variable = "CMCAT",
+#'   target_hardcoded_value = "GENERAL CONCOMITANT MEDICATIONS",
+#'   target_dataset = CM_INTER,
+#'   merge_to_topic_by = c("oak_id", "raw_source", "patient_number")
+#' )
+#'
+#' @importFrom rlang :=
+#' @export
+hardcode_no_ct <- function(raw_dataset,
+                           raw_variable,
+                           target_sdtm_variable,
+                           target_hardcoded_value,
+                           target_dataset = raw_dataset,
+                           merge_to_topic_by = NULL) {
+  dplyr::right_join(x = raw_dataset, y = target_dataset, by = merge_to_topic_by) |>
+    dplyr::mutate("{raw_variable}" := overwrite(!!rlang::sym(raw_variable), target_hardcoded_value)) |>
+    dplyr::rename("{target_sdtm_variable}" := raw_variable) |>
+    dplyr::relocate(target_sdtm_variable, .after = dplyr::last_col())
```

Comment:
```suggestion
  sdtm_ds <- raw_dataset |>
    dplyr::select(rlang::sym(raw_variable), merge_to_topic_by) |>
    dplyr::mutate(target_sdtm_variable = ifelse(!is.na(raw_dataset[[raw_variable]]),
                                                target_hardcoded_value,
                                                NA_character_)) |>
    dplyr::rename("{target_sdtm_variable}" := target_sdtm_variable)
  
  sdtm_ds <- dplyr::left_join(target_dataset, sdtm_ds, by = merge_to_topic_by) |>
    dplyr::select(-rlang::sym(raw_variable))
```

Reply:
This logic appears to be ok, but we should add more assertions to ensure its validity. Of course, there is always room for improvement.

Reply:
Hi Ramm,

Thank you for the feedback!

May I ask, if the logic is okay, why then the suggested changes? They seem to yield equivalent code to the one I already have, except for one difference: 

- when no argument is passed to `target_dataset`, then it defaults to `raw_dataset`, and with your suggestions that means you will drop all variables from the target dataset except for the join keys (`merge_to_topic_by`) and the variable to be mutated, i.e. `target_sdtm_variable`. Is this intended? I think not.

Moreover, in my code I use this helper function `overwrite()` which nicely wraps that if-else pattern to preserve `NA` (in your suggestion) when hardcoding values. I guess this function may be useful more times as this feels like is going to be recurrent pattern. I've also written a kind of sibling function `rewrite()` which nicely recodes values (in the same spirit of `dplyr::recode()`/`dplyr::case_when()`), and that I expect to be useful when dealing with controlled terminology mappings.

One more question: except for the join keys, should the variables in the raw and target data sets be mutually exclusive? Should I assert for this possibility, and trigger an error if there is at least one variable in common between the two data sets?

Reply:
In roak sometimes we notice that the same record gets value from two different topics causing VAR.x and VAR.y creation. In this case we halt the process and message to the user. Hopefully with this nice modular approach we wont see this happening. Perhaps having an assertion won't hurt.

- https://github.com/pharmaverse/roak_pilot/blob/main/R/oak_coalesce_non_joined_duplicate.R

Reply:
hi @edgar-manukyan, could you check whether this issue still applies with the most recent code?

Reply:
I think this is outdated now.